diff --git a/rpc/rpc.go b/rpc/rpc.go
index 96137bb77be5d6df69124f190f3b9d06ab029b92..9d4c93325fafd31e6966ee53064ffdf0d92bd910 100644
--- a/rpc/rpc.go
+++ b/rpc/rpc.go
@@ -1,40 +1,40 @@
 package rpc
 
 import (
 	"bytes"
 	"crypto/sha256"
 	"encoding/json"
 	"errors"
 	"fmt"
 	"math/big"
 	"net/http"
 	"strconv"
 	"strings"
 	"sync"
 
-	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
 
 	"github.com/sammy007/open-ethereum-pool/util"
 )
 
 type RPCClient struct {
 	sync.RWMutex
 	Url         string
 	Name        string
 	sick        bool
 	sickRate    int
 	successRate int
 	client      *http.Client
 }
 
 type GetBlockReply struct {
 	Number       string   `json:"number"`
 	Hash         string   `json:"hash"`
 	Nonce        string   `json:"nonce"`
 	Miner        string   `json:"miner"`
 	Difficulty   string   `json:"difficulty"`
 	GasLimit     string   `json:"gasLimit"`
 	GasUsed      string   `json:"gasUsed"`
 	Transactions []Tx     `json:"transactions"`
 	Uncles       []string `json:"uncles"`
 	// https://github.com/ethereum/EIPs/issues/95
@@ -66,50 +66,58 @@ func (r *TxReceipt) Successful() bool {
 	}
 	return true
 }
 
 type Tx struct {
 	Gas      string `json:"gas"`
 	GasPrice string `json:"gasPrice"`
 	Hash     string `json:"hash"`
 }
 
 type JSONRpcResp struct {
 	Id     *json.RawMessage       `json:"id"`
 	Result *json.RawMessage       `json:"result"`
 	Error  map[string]interface{} `json:"error"`
 }
 
 func NewRPCClient(name, url, timeout string) *RPCClient {
 	rpcClient := &RPCClient{Name: name, Url: url}
 	timeoutIntv := util.MustParseDuration(timeout)
 	rpcClient.client = &http.Client{
 		Timeout: timeoutIntv,
 	}
 	return rpcClient
 }
 
+func (r *RPCClient) Call(result interface{}, method string, params ...interface{}) error {
+	rpcResp, err := r.doPost(r.Url, method, params)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(*rpcResp.Result, result)
+}
+
 func (r *RPCClient) GetWork() ([]string, error) {
 	rpcResp, err := r.doPost(r.Url, "eth_getWork", []string{})
 	if err != nil {
 		return nil, err
 	}
 	var reply []string
 	err = json.Unmarshal(*rpcResp.Result, &reply)
 	return reply, err
 }
 
 func (r *RPCClient) GetPendingBlock() (*GetBlockReplyPart, error) {
 	rpcResp, err := r.doPost(r.Url, "eth_getBlockByNumber", []interface{}{"pending", false})
 	if err != nil {
 		return nil, err
 	}
 	if rpcResp.Result != nil {
 		var reply *GetBlockReplyPart
 		err = json.Unmarshal(*rpcResp.Result, &reply)
 		return reply, err
 	}
 	return nil, nil
 }
 
 func (r *RPCClient) GetBlockByHeight(height int64) (*GetBlockReply, error) {
 	params := []interface{}{fmt.Sprintf("0x%x", height), true}
@@ -155,51 +163,51 @@ func (r *RPCClient) GetTxReceipt(hash string) (*TxReceipt, error) {
 func (r *RPCClient) SubmitBlock(params []string) (bool, error) {
 	rpcResp, err := r.doPost(r.Url, "eth_submitWork", params)
 	if err != nil {
 		return false, err
 	}
 	var reply bool
 	err = json.Unmarshal(*rpcResp.Result, &reply)
 	return reply, err
 }
 
 func (r *RPCClient) GetBalance(address string) (*big.Int, error) {
 	rpcResp, err := r.doPost(r.Url, "eth_getBalance", []string{address, "latest"})
 	if err != nil {
 		return nil, err
 	}
 	var reply string
 	err = json.Unmarshal(*rpcResp.Result, &reply)
 	if err != nil {
 		return nil, err
 	}
 	return util.String2Big(reply), err
 }
 
 func (r *RPCClient) Sign(from string, s string) (string, error) {
 	hash := sha256.Sum256([]byte(s))
-	rpcResp, err := r.doPost(r.Url, "eth_sign", []string{from, common.ToHex(hash[:])})
+	rpcResp, err := r.doPost(r.Url, "eth_sign", []string{from, hexutil.Encode(hash[:])})
 	var reply string
 	if err != nil {
 		return reply, err
 	}
 	err = json.Unmarshal(*rpcResp.Result, &reply)
 	if err != nil {
 		return reply, err
 	}
 	if util.IsZeroHash(reply) {
 		err = errors.New("Can't sign message, perhaps account is locked")
 	}
 	return reply, err
 }
 
 func (r *RPCClient) GetPeerCount() (int64, error) {
 	rpcResp, err := r.doPost(r.Url, "net_peerCount", nil)
 	if err != nil {
 		return 0, err
 	}
 	var reply string
 	err = json.Unmarshal(*rpcResp.Result, &reply)
 	if err != nil {
 		return 0, err
 	}
 	return strconv.ParseInt(strings.Replace(reply, "0x", "", -1), 16, 64)

